src/python/dxpy/bindings/dxfile_functions.py

```diff
@@ -537,29 +497,87 @@ def report_progress(handler, num_bytes):
+                else:
+                    creation_kwargs['name'] = os.path.basename(local_file_name)
 
-    while True:
-        buf = read(handler._write_bufsize)
-        offset += len(buf)
+        # Use 'a' mode because we will be responsible for closing the file
+        # ourselves later (if requested).
+        return new_dxfile(mode='a', media_type=media_type, write_buffer_size=write_buffer_size,
+                             expected_file_size=file_size, file_is_mmapd=file_is_mmapd, **creation_kwargs)
 
-        if len(buf) == 0:
-            break
+    retries = 0
+    max_retries = 2
+    file_is_mmapd = hasattr(fd, "fileno")
 
-        handler.write(buf,
-                      report_progress_fn=report_progress if show_progress else None,
-                      multithread=multithread,
-                      **remaining_kwargs)
+    if write_buffer_size is None:
+        write_buffer_size=dxfile.DEFAULT_BUFFER_SIZE
 
-    if filename is not None:
-        fd.close()
+    # APPS-650 file upload would occasionally fail due to some parts not being uploaded correctly. This will try to re-upload in case this happens.
+    while retries <= max_retries:
+        retries += 1
 
-    handler.flush(report_progress_fn=report_progress if show_progress else None, **remaining_kwargs)
+        if use_existing_dxfile:
+            handler = use_existing_dxfile
+        else:
+            handler = get_new_handler(filename)
 
-    if show_progress:
-        sys.stderr.write("\n")
-        sys.stderr.flush()
+        # For subsequent API calls, don't supply the dataobject metadata
+        # parameters that are only needed at creation time.
+        _, remaining_kwargs = dxpy.DXDataObject._get_creation_params(kwargs)
+
+        num_ticks = 60
+        offset = 0
+
+        handler._ensure_write_bufsize(**remaining_kwargs)
+
+        handler._num_bytes_transmitted = 0
+
+        if show_progress:
+            report_progress(handler, 0)
+
+        while True:
+            buf = read(handler._write_bufsize)
+            offset += len(buf)
+
+            if len(buf) == 0:
+                break
+
+            handler.write(buf,
+                          report_progress_fn=report_progress if show_progress else None,
+                          multithread=multithread,
+                          **remaining_kwargs)
+
+        handler.flush(report_progress_fn=report_progress if show_progress else None, **remaining_kwargs)
+
+        if show_progress:
+            sys.stderr.write("\n")
+            sys.stderr.flush()
+        try:
+            handler.wait_until_parts_uploaded()
+        except DXError:
+            if show_progress:
+                logger.warning("File {} was not uploaded correctly!".format(filename))
+            if retries > max_retries:
+                raise
+            if show_progress:
+                logger.warning("Retrying...({}/{})".format(retries, max_retries))
+            continue
+        if filename is not None:
```

Comment:
how does this work when uploading from a stding fife (e.g. `dx upload --path x -`)?

Reply:
I tried `dx upload /dev/stdin` which worked. I guess this is the same thing, right?